std-bits/table/src/main/java/org/enso/table/write/ExcelWriter.java

```diff
@@ -127,47 +126,99 @@ public static void writeTableToRange(Workbook workbook, ExcelRange range, boolea
-            ? ExcelHeaders.HeaderBehavior.USE_FIRST_ROW_AS_HEADERS
-            : ExcelHeaders.HeaderBehavior.EXCEL_COLUMN_NAMES;
-      }
+      expanded = new ExcelRange(expanded.getSheetName(), expanded.getLeftColumn(), finalRow + 1, expanded.getRightColumn(), expanded.getBottomRow());
+    }
+
+    updateRangeWithTable(workbook, expanded, false, existingDataMode, mappedTable, rowLimit, ExcelHeaders.HeaderBehavior.EXCEL_COLUMN_NAMES, sheet);
+  }
+
+  private static void updateRangeWithTable(Workbook workbook, ExcelRange range, boolean singleCell, ExistingDataMode existingDataMode, Table table, Long rowLimit, ExcelHeaders.HeaderBehavior headers, ExcelSheet sheet)
+      throws RangeExceededException, ExistingDataException {
+    boolean writeHeaders = headers == ExcelHeaders.HeaderBehavior.USE_FIRST_ROW_AS_HEADERS;
+    int requiredRows = Math.min(table.rowCount(), rowLimit == null ? Integer.MAX_VALUE : rowLimit.intValue()) + (writeHeaders ? 1 : 0);
+
+    // Expand Single Cell if needed.
+    if (singleCell) {
+      range = new ExcelRange(
+          range.getSheetName(),
+          range.getLeftColumn(),
+          range.getTopRow(),
+          Math.max(range.getRightColumn(), range.getLeftColumn() + table.getColumns().length - 1),
+          Math.max(range.getBottomRow(), range.getTopRow() + requiredRows - 1));
+    }
+
+    // Check Size of Range
+    int finalRow = range.isWholeColumn() ? workbook.getSpreadsheetVersion().getMaxRows() : range.getBottomRow();
+    int availableRows = finalRow - range.getTopRow() + 1;
+    if (range.getColumnCount() < table.getColumns().length || availableRows < requiredRows) {
+      throw new RangeExceededException("Range is too small to fit all data.");
+    }
 
-      checkExistingRange(workbook, range, replace, sheet);
+    // Check or Clear Current Range
+    if (rangeIsNotEmpty(workbook, range, existingDataMode == ExistingDataMode.REPLACE, sheet)) {
+      throw new ExistingDataException("Range is not empty, and cannot be replaced in current mode.");
     }
 
-    writeTableToSheet(workbook, sheet.getSheet(), range.getTopRow() - 1, range.getLeftColumn(), table, rowLimit, headers != ExcelHeaders.HeaderBehavior.EXCEL_COLUMN_NAMES);
+    writeTableToSheet(workbook, sheet.getSheet(), range.getTopRow() - 1, range.getLeftColumn(), table, rowLimit, writeHeaders);
   }
 
-  private static void checkExistingRange(Workbook workbook, ExcelRange range, boolean replace, ExcelSheet sheet) throws ExistingDataException {
+  /***
+   * Checks if a range is empty.
+   * @param workbook The workbook to check.
+   * @param range The range to check.
+   * @param clear Clear the cells if they are not empty.
+   * @param sheet Sheet containing the range.
+   * @return True if range is empty and clear is False, otherwise returns False.
+   */
+  private static boolean rangeIsNotEmpty(Workbook workbook, ExcelRange range, boolean clear, ExcelSheet sheet) {
```

Comment:
The name of the method and the top doc-comment suggest that this is a read-only method checking something. Only after checking the doc for the `clear` argument we see that it actually performs side-effects.

I'd suggest making this more clear in the top comment and possibly renaming the method to reflect that it performs side-effects. Does this have to be a single method in the first place? Seems to be breaking SRP a bit and I'm not sure if that is necessary. It may be a subjective thing, but I feel like splitting this into two methods like: `checkRangeIsEmpty` and `clearRange` would lead to more readable code. The `topRow` etc. logic may be split to some helper function to avoid duplicating it then.